README
Chronic
http://chronic.rubyforge.org/
	by Tom Preston-Werner

== DESCRIPTION:

Chronic is a natural language date/time parser written in pure Ruby. See below for the wide variety of formats Chronic will parse.

== INSTALLATION:

Chronic can be installed via RubyGems:

  $ sudo gem install chronic
  
== CODE:

Browse the code and get an RSS feed of the commit log at:

  http://github.com/mojombo/chronic.git

You can grab the code (and help with development) via git:

  $ git clone git://github.com/mojombo/chronic.git

== USAGE:

You can parse strings containing a natural language date using the Chronic.parse method.

  require 'rubygems'
  require 'chronic'

  Time.now   #=> Sun Aug 27 23:18:25 PDT 2006

  #---

  Chronic.parse('tomorrow')       
    #=> Mon Aug 28 12:00:00 PDT 2006

  Chronic.parse('monday', :context => :past)
    #=> Mon Aug 21 12:00:00 PDT 2006

  Chronic.parse('this tuesday 5:00')
    #=> Tue Aug 29 17:00:00 PDT 2006

  Chronic.parse('this tuesday 5:00', :ambiguous_time_range => :none)
    #=> Tue Aug 29 05:00:00 PDT 2006

  Chronic.parse('may 27th', :now => Time.local(2000, 1, 1))
    #=> Sat May 27 12:00:00 PDT 2000

  Chronic.parse('may 27th', :guess => false)
    #=> Sun May 27 00:00:00 PDT 2007..Mon May 28 00:00:00 PDT 2007

See Chronic.parse for detailed usage instructions.

== EXAMPLES:

Chronic can parse a huge variety of date and time formats. Following is a small sample of strings that will be properly parsed. Parsing is case insensitive and will handle common abbreviations and misspellings.

Simple

  thursday
  november
  summer
  friday 13:00
  mon 2:35
  4pm
  6 in the morning
  friday 1pm
  sat 7 in the evening
  yesterday
  today
  tomorrow
  this tuesday
  next month
  last winter
  this morning
  last night
  this second
  yesterday at 4:00
  last friday at 20:00
  last week tuesday
  tomorrow at 6:45pm
  afternoon yesterday
  thursday last week

Complex

  3 years ago
  5 months before now
  7 hours ago
  7 days from now
  1 week hence
  in 3 hours
  1 year ago tomorrow
  3 months ago saturday at 5:00 pm
  7 hours before tomorrow at noon
  3rd wednesday in november
  3rd month next year
  3rd thursday this september
  4th day last week

Specific Dates

  January 5
  dec 25
  may 27th
  October 2006
  oct 06
  jan 3 2010
  february 14, 2004
  3 jan 2000
  17 april 85
  5/27/1979
  27/5/1979
  05/06
  1979-05-27
  Friday
  5
  4:00
  17:00
  0800

Specific Times (many of the above with an added time)

  January 5 at 7pm
  1979-05-27 05:00:00
  etc

== TIME ZONES:

Chronic allows you to set which Time class to use when constructing times.  By default, the built in Ruby time class creates times in your system's
local time zone.  You can set this to something like ActiveSupport's TimeZone class to get full time zone support.

  >> Time.zone = "UTC"
  >> Chronic.time_class = Time.zone
  >> Chronic.parse("June 15 2006 at 5:45 AM")
  => Thu, 15 Jun 2006 05:45:00 UTC +00:00

== LIMITATIONS:
  
Chronic uses Ruby's built in Time class for all time storage and computation. Because of this, only times that the Time class can handle will be properly parsed. Parsing for times outside of this range will simply return nil. Support for a wider range of times is planned for a future release.

== LICENSE:

(The MIT License)

Copyright (c) 2008 Tom Preston-Werner

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
= 0.3.9   2010-08-04

* fixed 2 digit dates in Ruby 1.9.2
* fixed timezones in Ruby 1.9.2
* fixed Tick addition error in Ruby 1.9.2
* passing tests in Ruby 1.8.7 and 1.9.2
* support for periods as separators in time
* update gemspec name to aaronh-chronic for publish to rubygems.org

= 0.2.4

* fixed numerizer number combination bug (27 Oct 2006 7:30pm works now) (reported by reynard h)
* allow numeric timezone offset (e.g -0500)
* disregard commas (so as to not return nil)
* fix parse of (am|pm|oclock) separation to handle "Ham sandwich" properly
* handle 'on' e.g. 5pm on Monday (by indirect)
* support seasons (by Edwin Chen)
* support weekend/weekday (by Edwin Chen)
* add endianness option (by Mark Johnson/Thomas Lee)
* updated version number in the module (Jeffrey 'jf' Lim)
* fixed/improved logic checks in Ordinal, and Scalar (Jeffrey 'jf' Lim)

= 0.2.3

* fixed 12am/12pm (by Nicholas Schlueter)

= 0.2.2

* added missing files (damn you manifest)

= 0.2.1

* fixed time overflow issue
* implemented "next" for minute repeater
* generalized time dealiasing to dealias regardless of day portion and time position
* added additional token match for cases like "friday evening at 7" and "tomorrow evening at 7"
* added support for Time#to_s output format: "Mon Apr 02 17:00:00 PDT 2007"

= 0.2.0 2007-03-20

* implemented numerizer, allowing the use of number words (e.g. five weeks ago) (by shalev)

= 0.1.6 2006-01-15

* added 'weekend' support (by eventualbuddha)

= 0.1.5 2006-12-20

* fixed 'aug 20' returning next year if current month is august
* modified behavior of 'from now'
* added support for seconds on times, and thus db timestamp format: "2006-12-20 18:04:23"
* made Hoe compliant

= 0.1.4

* removed verbose error checking code. oops. :-/

= 0.1.3

* improved regexes for word variations (by Josh Goebel)
* fixed a bug that caused "today at 3am" to return nil if current time is after 3am

= 0.1.2

* removed Date dependency (now works on windows properly without fiddling)

= 0.1.1

* run to_s on incoming object
* fixed loop loading of repeaters files (out of order on some machines)
* fixed find_within to use this instead of next (was breaking "today at 6pm")

= 0.1.0

* initial release


lib/chronic/handlers.rb
module Chronic

	class << self
	  
    def definitions(options={}) #:nodoc:
      options[:endian_precedence] = [:middle, :little] if options[:endian_precedence].nil?
      
      # ensure the endian precedence is exactly two elements long
      raise ChronicPain, "More than two elements specified for endian precedence array" unless options[:endian_precedence].length == 2

      # handler for dd/mm/yyyy
      @little_endian_handler ||= Handler.new([:scalar_day, :separator_slash_or_dash, :scalar_month, :separator_slash_or_dash, :scalar_year, :separator_at?, 'time?'], :handle_sd_sm_sy)

      # handler for mm/dd/yyyy
      @middle_endian_handler ||= Handler.new([:scalar_month, :separator_slash_or_dash, :scalar_day, :separator_slash_or_dash, :scalar_year, :separator_at?, 'time?'], :handle_sm_sd_sy)

      # ensure we have valid endian values
      options[:endian_precedence].each do |e|
        raise ChronicPain, "Unknown endian type: #{e.to_s}" unless instance_variable_defined?(endian_variable_name_for(e))
      end

	    @definitions ||= 
      {:time => [Handler.new([:repeater_time, :repeater_day_portion?], nil)],
        
       :date => [Handler.new([:repeater_day_name, :repeater_month_name, :scalar_day, :repeater_time, :separator_slash_or_dash?, :time_zone, :scalar_year], :handle_rdn_rmn_sd_t_tz_sy),
                 Handler.new([:repeater_month_name, :scalar_day, :scalar_year], :handle_rmn_sd_sy),
                 Handler.new([:repeater_month_name, :scalar_day, :scalar_year, :separator_at?, 'time?'], :handle_rmn_sd_sy),
                 Handler.new([:repeater_month_name, :scalar_day, :separator_at?, 'time?'], :handle_rmn_sd),
                 Handler.new([:repeater_time, :repeater_day_portion?, :separator_on?, :repeater_month_name, :scalar_day], :handle_rmn_sd_on),
                 Handler.new([:repeater_month_name, :ordinal_day, :separator_at?, 'time?'], :handle_rmn_od),
                 Handler.new([:repeater_time, :repeater_day_portion?, :separator_on?, :repeater_month_name, :ordinal_day], :handle_rmn_od_on),
                 Handler.new([:repeater_month_name, :scalar_year], :handle_rmn_sy),
                 Handler.new([:scalar_day, :repeater_month_name, :scalar_year, :separator_at?, 'time?'], :handle_sd_rmn_sy),
                 @middle_endian_handler,
                 @little_endian_handler,
                 Handler.new([:scalar_year, :separator_slash_or_dash, :scalar_month, :separator_slash_or_dash, :scalar_day, :separator_at?, 'time?'], :handle_sy_sm_sd),
                 Handler.new([:scalar_month, :separator_slash_or_dash, :scalar_year], :handle_sm_sy)],
                 
       # tonight at 7pm
       :anchor => [Handler.new([:grabber?, :repeater, :separator_at?, :repeater?, :repeater?], :handle_r),
                   Handler.new([:grabber?, :repeater, :repeater, :separator_at?, :repeater?, :repeater?], :handle_r),
                   Handler.new([:repeater, :grabber, :repeater], :handle_r_g_r)],
                   
       # 3 weeks from now, in 2 months
       :arrow => [Handler.new([:scalar, :repeater, :pointer], :handle_s_r_p),
                  Handler.new([:pointer, :scalar, :repeater], :handle_p_s_r),
                  Handler.new([:scalar, :repeater, :pointer, 'anchor'], :handle_s_r_p_a)],
                  
       # 3rd week in march
       :narrow => [Handler.new([:ordinal, :repeater, :separator_in, :repeater], :handle_o_r_s_r),
                   Handler.new([:ordinal, :repeater, :grabber, :repeater], :handle_o_r_g_r)]
      }
      
      apply_endian_precedences(options[:endian_precedence])
      
      @definitions
    end
    
    def tokens_to_span(tokens, options) #:nodoc:
      # maybe it's a specific date
      
      definitions = self.definitions(options)
      definitions[:date].each do |handler|
        if handler.match(tokens, definitions)
          puts "-date" if Chronic.debug
          good_tokens = tokens.select { |o| !o.get_tag Separator }
          return self.send(handler.handler_method, good_tokens, options)
        end
      end
            
      # I guess it's not a specific date, maybe it's just an anchor
            
      definitions[:anchor].each do |handler|
        if handler.match(tokens, definitions)
          puts "-anchor" if Chronic.debug
          good_tokens = tokens.select { |o| !o.get_tag Separator }
          return self.send(handler.handler_method, good_tokens, options)
        end
      end
            
      # not an anchor, perhaps it's an arrow
      
      definitions[:arrow].each do |handler|
        if handler.match(tokens, definitions)
          puts "-arrow" if Chronic.debug
          good_tokens = tokens.reject { |o| o.get_tag(SeparatorAt) || o.get_tag(SeparatorSlashOrDash) || o.get_tag(SeparatorComma) }
          return self.send(handler.handler_method, good_tokens, options)
        end
      end
      
      # not an arrow, let's hope it's a narrow
      
      definitions[:narrow].each do |handler|
        if handler.match(tokens, definitions)
          puts "-narrow" if Chronic.debug
          #good_tokens = tokens.select { |o| !o.get_tag Separator }
          return self.send(handler.handler_method, tokens, options)
        end
      end
      
      # I guess you're out of luck!
      puts "-none" if Chronic.debug
      return nil
    end
    
    #--------------
    
    def apply_endian_precedences(precedences)
      date_defs = @definitions[:date]

      # map the precedence array to indices on @definitions[:date]
      indices = precedences.map { |e|
        handler = instance_variable_get(endian_variable_name_for(e))
        date_defs.index(handler)
      }

      # swap the handlers if we discover they are at odds with the desired preferences
      swap(date_defs, indices.first, indices.last) if indices.first > indices.last
    end

    def endian_variable_name_for(e)
      "@#{e.to_s}_endian_handler".to_sym
    end

    # exchange two elements in an array
    def swap(arr, a, b); arr[a], arr[b] = arr[b], arr[a]; end

    def day_or_time(day_start, time_tokens, options)
      outer_span = Span.new(day_start, day_start + (24 * 60 * 60))
      
      if !time_tokens.empty?
        @now = outer_span.begin
        time = get_anchor(dealias_and_disambiguate_times(time_tokens, options), options)
        return time
      else
        return outer_span
      end
    end
    
    #--------------
    
    def handle_m_d(month, day, time_tokens, options) #:nodoc:
      month.start = @now
      span = month.this(options[:context])
      
      day_start = Chronic.time_class.local(span.begin.year, span.begin.month, day)
      
      day_or_time(day_start, time_tokens, options)
    end
    
    def handle_rmn_sd(tokens, options) #:nodoc:
      handle_m_d(tokens[0].get_tag(RepeaterMonthName), tokens[1].get_tag(ScalarDay).type, tokens[2..tokens.size], options)
    end
    
    def handle_rmn_sd_on(tokens, options) #:nodoc:
      if tokens.size > 3
        handle_m_d(tokens[2].get_tag(RepeaterMonthName), tokens[3].get_tag(ScalarDay).type, tokens[0..1], options)
      else
        handle_m_d(tokens[1].get_tag(RepeaterMonthName), tokens[2].get_tag(ScalarDay).type, tokens[0..0], options)
      end
    end
    
    def handle_rmn_od(tokens, options) #:nodoc:
      handle_m_d(tokens[0].get_tag(RepeaterMonthName), tokens[1].get_tag(OrdinalDay).type, tokens[2..tokens.size], options)
    end

    def handle_rmn_od_on(tokens, options) #:nodoc:
      if tokens.size > 3
        handle_m_d(tokens[2].get_tag(RepeaterMonthName), tokens[3].get_tag(OrdinalDay).type, tokens[0..1], options)
      else
        handle_m_d(tokens[1].get_tag(RepeaterMonthName), tokens[2].get_tag(OrdinalDay).type, tokens[0..0], options)
      end
    end
    
    def handle_rmn_sy(tokens, options) #:nodoc:
      month = tokens[0].get_tag(RepeaterMonthName).index
      year = tokens[1].get_tag(ScalarYear).type
      
      if month == 12
        next_month_year = year + 1
        next_month_month = 1
      else
        next_month_year = year
        next_month_month = month + 1
      end
      
      begin
        Span.new(Chronic.time_class.local(year, month), Chronic.time_class.local(next_month_year, next_month_month))
      rescue ArgumentError
        nil
      end
    end
    
    def handle_rdn_rmn_sd_t_tz_sy(tokens, options) #:nodoc:
      t = Chronic.time_class.parse(@text)
      Span.new(t, t + 1)
    end
    
    def handle_rmn_sd_sy(tokens, options) #:nodoc:
      month = tokens[0].get_tag(RepeaterMonthName).index
      day = tokens[1].get_tag(ScalarDay).type
      year = tokens[2].get_tag(ScalarYear).type
      
      time_tokens = tokens.last(tokens.size - 3)
      
      begin
        day_start = Chronic.time_class.local(year, month, day)
        day_or_time(day_start, time_tokens, options)
      rescue ArgumentError
        nil
      end
    end
    
    def handle_sd_rmn_sy(tokens, options) #:nodoc:
      new_tokens = [tokens[1], tokens[0], tokens[2]]
      time_tokens = tokens.last(tokens.size - 3)
      self.handle_rmn_sd_sy(new_tokens + time_tokens, options)
    end
    
    def handle_sm_sd_sy(tokens, options) #:nodoc:
      month = tokens[0].get_tag(ScalarMonth).type
      day = tokens[1].get_tag(ScalarDay).type
      year = tokens[2].get_tag(ScalarYear).type
      
      time_tokens = tokens.last(tokens.size - 3)
      
      begin
        day_start = Chronic.time_class.local(year, month, day) #:nodoc:
        day_or_time(day_start, time_tokens, options)
      rescue ArgumentError
        nil
      end
    end
    
    def handle_sd_sm_sy(tokens, options) #:nodoc:
      new_tokens = [tokens[1], tokens[0], tokens[2]]
      time_tokens = tokens.last(tokens.size - 3)
      self.handle_sm_sd_sy(new_tokens + time_tokens, options)
    end
    
    def handle_sy_sm_sd(tokens, options) #:nodoc:
      new_tokens = [tokens[1], tokens[2], tokens[0]]
      time_tokens = tokens.last(tokens.size - 3)
      self.handle_sm_sd_sy(new_tokens + time_tokens, options)
    end
    
    def handle_sm_sy(tokens, options) #:nodoc:
      month = tokens[0].get_tag(ScalarMonth).type
      year = tokens[1].get_tag(ScalarYear).type
      
      if month == 12
        next_month_year = year + 1
        next_month_month = 1
      else
        next_month_year = year
        next_month_month = month + 1
      end
      
      begin
        Span.new(Chronic.time_class.local(year, month), Chronic.time_class.local(next_month_year, next_month_month))
      rescue ArgumentError
        nil
      end
    end
    
    # anchors
    
    def handle_r(tokens, options) #:nodoc:
      dd_tokens = dealias_and_disambiguate_times(tokens, options)
      self.get_anchor(dd_tokens, options)
    end
    
    def handle_r_g_r(tokens, options) #:nodoc:
      new_tokens = [tokens[1], tokens[0], tokens[2]]
      self.handle_r(new_tokens, options)
    end
    
    # arrows
    
    def handle_srp(tokens, span, options) #:nodoc:
      distance = tokens[0].get_tag(Scalar).type
      repeater = tokens[1].get_tag(Repeater)
      pointer = tokens[2].get_tag(Pointer).type
      
      repeater.offset(span, distance, pointer)
    end
    
    def handle_s_r_p(tokens, options) #:nodoc:
      repeater = tokens[1].get_tag(Repeater)
            
      # span = 
      # case true
      # when [RepeaterYear, RepeaterSeason, RepeaterSeasonName, RepeaterMonth, RepeaterMonthName, RepeaterFortnight, RepeaterWeek].include?(repeater.class)
      #   self.parse("this hour", :guess => false, :now => @now)
      # when [RepeaterWeekend, RepeaterDay, RepeaterDayName, RepeaterDayPortion, RepeaterHour].include?(repeater.class)
      #   self.parse("this minute", :guess => false, :now => @now)
      # when [RepeaterMinute, RepeaterSecond].include?(repeater.class)
      #   self.parse("this second", :guess => false, :now => @now)
      # else
      #   raise(ChronicPain, "Invalid repeater: #{repeater.class}")
      # end
      
      span = self.parse("this second", :guess => false, :now => @now)
      
      self.handle_srp(tokens, span, options)
    end
    
    def handle_p_s_r(tokens, options) #:nodoc:
      new_tokens = [tokens[1], tokens[2], tokens[0]]
      self.handle_s_r_p(new_tokens, options)
    end
    
    def handle_s_r_p_a(tokens, options) #:nodoc:
      anchor_span = get_anchor(tokens[3..tokens.size - 1], options)
      self.handle_srp(tokens, anchor_span, options)
    end
    
    # narrows
    
    def handle_orr(tokens, outer_span, options) #:nodoc:
      repeater = tokens[1].get_tag(Repeater)
      repeater.start = outer_span.begin - 1
      ordinal = tokens[0].get_tag(Ordinal).type
      span = nil
      ordinal.times do
        span = repeater.next(:future)
        if span.begin > outer_span.end
          span = nil
          break
        end
      end
      span
    end
    
    def handle_o_r_s_r(tokens, options) #:nodoc:
      outer_span = get_anchor([tokens[3]], options)
      handle_orr(tokens[0..1], outer_span, options)
    end
    
    def handle_o_r_g_r(tokens, options) #:nodoc:
      outer_span = get_anchor(tokens[2..3], options)
      handle_orr(tokens[0..1], outer_span, options)
    end
    
    # support methods
    
    def get_anchor(tokens, options) #:nodoc:
      grabber = Grabber.new(:this)
      pointer = :future
      
      repeaters = self.get_repeaters(tokens)
      repeaters.size.times { tokens.pop }
                    
      if tokens.first && tokens.first.get_tag(Grabber)
        grabber = tokens.first.get_tag(Grabber)
        tokens.pop
      end
      
      head = repeaters.shift
      head.start = @now
                  
      case grabber.type
        when :last
          outer_span = head.next(:past)
        when :this
          if repeaters.size > 0
            outer_span = head.this(:none)
          else
            outer_span = head.this(options[:context])
          end
        when :next
          outer_span = head.next(:future)
        else raise(ChronicPain, "Invalid grabber")
      end
      
      puts "--#{outer_span}" if Chronic.debug
      anchor = find_within(repeaters, outer_span, pointer)
    end
    
    def get_repeaters(tokens) #:nodoc:
      repeaters = []
		  tokens.each do |token|
		    if t = token.get_tag(Repeater)
          repeaters << t
        end
      end
      repeaters.sort.reverse
    end
    
    def in_span?(span, t)
      t && span && t >= span.begin && t <= span.end
    end
    
    # Recursively finds repeaters within other repeaters.
    # Returns a Span representing the innermost time span
    # or nil if no repeater union could be found
    def find_within(tags, span, pointer) #:nodoc:
      puts "--#{span}" if Chronic.debug
      return span if tags.empty?
      
      head, *rest = tags
      head.start = pointer == :future ? span.begin : span.end
      h = head.this(:none)
            
      if in_span?(span, h.begin) || in_span?(span, h.end)
        return find_within(rest, h, pointer)
      else
        return nil
      end
    end
    
    def dealias_and_disambiguate_times(tokens, options) #:nodoc:
      # handle aliases of am/pm
      # 5:00 in the morning -> 5:00 am
      # 7:00 in the evening -> 7:00 pm
      
      day_portion_index = nil
      tokens.each_with_index do |t, i|
        if t.get_tag(RepeaterDayPortion)
          day_portion_index = i
          break
        end
      end
       
      time_index = nil
      tokens.each_with_index do |t, i|
        if t.get_tag(RepeaterTime)
          time_index = i
          break
        end
      end
      
      if (day_portion_index && time_index)
        t1 = tokens[day_portion_index]
        t1tag = t1.get_tag(RepeaterDayPortion)
      
        if [:morning].include?(t1tag.type)
          puts '--morning->am' if Chronic.debug
          t1.untag(RepeaterDayPortion)
          t1.tag(RepeaterDayPortion.new(:am))
        elsif [:afternoon, :evening, :night].include?(t1tag.type)
          puts "--#{t1tag.type}->pm" if Chronic.debug
          t1.untag(RepeaterDayPortion)
          t1.tag(RepeaterDayPortion.new(:pm))
        end
      end
      
      # tokens.each_with_index do |t0, i|
      #   t1 = tokens[i + 1]
      #   if t1 && (t1tag = t1.get_tag(RepeaterDayPortion)) && t0.get_tag(RepeaterTime)
      #     if [:morning].include?(t1tag.type)
      #       puts '--morning->am' if Chronic.debug
      #       t1.untag(RepeaterDayPortion)
      #       t1.tag(RepeaterDayPortion.new(:am))
      #     elsif [:afternoon, :evening, :night].include?(t1tag.type)
      #       puts "--#{t1tag.type}->pm" if Chronic.debug
      #       t1.untag(RepeaterDayPortion)
      #       t1.tag(RepeaterDayPortion.new(:pm))
      #     end
      #   end
      # end
            
      # handle ambiguous times if :ambiguous_time_range is specified
      if options[:ambiguous_time_range] != :none
        ttokens = []
        tokens.each_with_index do |t0, i|
          ttokens << t0
          t1 = tokens[i + 1]
          if t0.get_tag(RepeaterTime) && t0.get_tag(RepeaterTime).type.ambiguous? && (!t1 || !t1.get_tag(RepeaterDayPortion))
            distoken = Token.new('disambiguator')
            distoken.tag(RepeaterDayPortion.new(options[:ambiguous_time_range]))
            ttokens << distoken
          end
        end
        tokens = ttokens
      end
      
      tokens
    end
    
  end
  
  class Handler #:nodoc:
    attr_accessor :pattern, :handler_method
    
    def initialize(pattern, handler_method)
      @pattern = pattern
      @handler_method = handler_method
    end
    
    def constantize(name)
      camel = name.to_s.gsub(/(^|_)(.)/) { $2.upcase }
      ::Chronic.module_eval(camel, __FILE__, __LINE__)
    end
    
    def match(tokens, definitions)
      token_index = 0
      @pattern.each do |element|
        name = element.to_s
        optional = name.reverse[0..0] == '?'
        name = name.chop if optional
        if element.instance_of? Symbol
          klass = constantize(name)
          match = tokens[token_index] && !tokens[token_index].tags.select { |o| o.kind_of?(klass) }.empty?
          return false if !match && !optional
          (token_index += 1; next) if match
          next if !match && optional
        elsif element.instance_of? String
          return true if optional && token_index == tokens.size
          sub_handlers = definitions[name.intern] || raise(ChronicPain, "Invalid subset #{name} specified")
          sub_handlers.each do |sub_handler|
            return true if sub_handler.match(tokens[token_index..tokens.size], definitions)
          end
          return false
        else
          raise(ChronicPain, "Invalid match type: #{element.class}")
        end
      end
      return false if token_index != tokens.size
      return true
    end

    def ==(other)
      self.pattern == other.pattern
    end
  end
  
end


lib/chronic/repeaters/repeater_month.rb
class Chronic::RepeaterMonth < Chronic::Repeater #:nodoc:
  MONTH_SECONDS = 2_592_000 # 30 * 24 * 60 * 60
  YEAR_MONTHS = 12

  def initialize(type)
    super
    @current_month_start = nil
  end

  def next(pointer)
    super
    
    if !@current_month_start
      @current_month_start = offset_by(Time.construct(@now.year, @now.month), 1, pointer)
    else
      @current_month_start = offset_by(Time.construct(@current_month_start.year, @current_month_start.month), 1, pointer)
    end
    
    Chronic::Span.new(@current_month_start, Time.construct(@current_month_start.year, @current_month_start.month + 1))
  end
  
  def this(pointer = :future)
    super
    
    case pointer
    when :future
      month_start = Time.construct(@now.year, @now.month, @now.day + 1)
      month_end = self.offset_by(Time.construct(@now.year, @now.month), 1, :future)
    when :past
      month_start = Time.construct(@now.year, @now.month)
      month_end = Time.construct(@now.year, @now.month, @now.day)
    when :none
      month_start = Time.construct(@now.year, @now.month)
      month_end = self.offset_by(Time.construct(@now.year, @now.month), 1, :future)
    end
    
    Chronic::Span.new(month_start, month_end)
  end
  
  def offset(span, amount, pointer)      
    Chronic::Span.new(offset_by(span.begin, amount, pointer), offset_by(span.end, amount, pointer))
  end
  
  def offset_by(time, amount, pointer) 
    direction = pointer == :future ? 1 : -1
    
    amount_years = direction * amount / YEAR_MONTHS
    amount_months = direction * amount % YEAR_MONTHS
    
    new_year = time.year + amount_years
    new_month = time.month + amount_months
    if new_month > YEAR_MONTHS
      new_year += 1
      new_month -= YEAR_MONTHS
    end
    
    new_time = Time.construct(new_year, new_month, time.day, time.hour, time.min, time.sec)
    if new_time.month > new_month
      d = Date.new(new_year, new_month, -1)
      Time.construct(new_year, d.month, d.day, time.hour, time.min, time.sec)
    else
      new_time
    end
  end
  
  def width
    MONTH_SECONDS
  end
  
  def to_s
    super << '-month'
  end
end

lib/chronic.rb
#=============================================================================
#
#  Name:       Chronic
#  Author:     Tom Preston-Werner
#  Purpose:    Parse natural language dates and times into Time or
#              Chronic::Span objects
#
#=============================================================================

$:.unshift File.dirname(__FILE__)     # For use/testing when no gem is installed

require 'time'

require 'chronic/chronic'
require 'chronic/handlers'

require 'chronic/repeater'
require 'chronic/repeaters/repeater_year'
require 'chronic/repeaters/repeater_season'
require 'chronic/repeaters/repeater_season_name'
require 'chronic/repeaters/repeater_month'
require 'chronic/repeaters/repeater_month_name'
require 'chronic/repeaters/repeater_fortnight'
require 'chronic/repeaters/repeater_week'
require 'chronic/repeaters/repeater_weekend'
require 'chronic/repeaters/repeater_weekday'
require 'chronic/repeaters/repeater_day'
require 'chronic/repeaters/repeater_day_name'
require 'chronic/repeaters/repeater_day_portion'
require 'chronic/repeaters/repeater_hour'
require 'chronic/repeaters/repeater_minute'
require 'chronic/repeaters/repeater_second'
require 'chronic/repeaters/repeater_time'

require 'chronic/grabber'
require 'chronic/pointer'
require 'chronic/scalar'
require 'chronic/ordinal'
require 'chronic/separator'
require 'chronic/time_zone'

require 'numerizer/numerizer'

module Chronic
  VERSION = "0.3.8"
  
  class << self
    attr_accessor :debug

    def time_class
      Thread.current[:chronic_time_class] ||= Time
    end

    def time_class=(klass)
      Thread.current[:chronic_time_class] = klass
    end
  end

  self.debug = false
end

# class Time
#   def self.construct(year, month = 1, day = 1, hour = 0, minute = 0, second = 0)
#     # extra_seconds = second > 60 ? second - 60 : 0
#     # extra_minutes = minute > 59 ? minute - 59 : 0
#     # extra_hours = hour > 23 ? hour - 23 : 0
#     # extra_days = day > 
#     
#     if month > 12
#       if month % 12 == 0
#         year += (month - 12) / 12
#         month = 12
#       else
#         year += month / 12
#         month = month % 12
#       end
#     end
#     
#     base = Time.local(year, month)
#     puts base
#     offset = ((day - 1) * 24 * 60 * 60) + (hour * 60 * 60) + (minute * 60) + second
#     puts offset.to_s
#     date = base + offset
#     puts date
#     date
#   end
# end

class Time
  def self.construct(year, month = 1, day = 1, hour = 0, minute = 0, second = 0)
    
    if second >= 60
      minute += second / 60
      second = second % 60
    end
    
    if minute >= 60
      hour += minute / 60
      minute = minute % 60
    end
    
    if hour >= 24
      day += hour / 24
      hour = hour % 24
    end
    
    # determine if there is a day overflow. this is complicated by our crappy calendar
    # system (non-constant number of days per month)
    day <= 56 || raise("day must be no more than 56 (makes month resolution easier)")
    if day > 28
      # no month ever has fewer than 28 days, so only do this if necessary
      leap_year = (year % 4 == 0) && !(year % 100 == 0)
      leap_year_month_days = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
      common_year_month_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
      days_this_month = leap_year ? leap_year_month_days[month - 1] : common_year_month_days[month - 1]
      if day > days_this_month
        month += day / days_this_month
        day = day % days_this_month
      end
    end
    
    if month > 12
      if month % 12 == 0
        year += (month - 12) / 12
        month = 12
      else
        year += month / 12
        month = month % 12
      end
    end
    
    Chronic.time_class.local(year, month, day, hour, minute, second)
  end
end
